File: frontend/src/app/modules/BenchMark.ts
import { ID } from "@datorama/akita";
import { TipoLocalidade } from "./enum/TipoLocalidade";
import { ComparacaoDTO } from "./dto/ComparacaoDTO";

export class BenchMark {
  id: ID;
  nome: string;
  localidade1: string;
  localidade2: string;
  estadoLocalidade1: string;
  estadoLocalidade2: string;
  dataInicial: Date;
  dataFinal: Date;
  comparacao: ComparacaoDTO[] = [];
  tipoLocalidade: TipoLocalidade;

  constructor(
    id: ID,
    nome: string = '',
    localidade1: string = '',
    localidade2: string = '',
    estadoLocalidade1: string = '',
    estadoLocalidade2: string = '',
    dataInicial: Date = new Date(),
    dataFinal: Date = new Date(),
    comparacao: ComparacaoDTO[] = [],
    tipoLocalidade: TipoLocalidade,
  ) {
    this.id = id;
    this.nome = nome;
    this.estadoLocalidade1 = estadoLocalidade1;
    this.estadoLocalidade2 = estadoLocalidade2;
    this.localidade1 = localidade1;
    this.localidade2 = localidade2;
    this.dataInicial = dataInicial;
    this.dataFinal = dataFinal;
    this.comparacao = comparacao;
    this.tipoLocalidade = tipoLocalidade;
  }

}

File: backend/src/main/java/com/backend/BenchMarks/model/BenchMark.java
package com.backend.BenchMarks.model;

import java.time.LocalDate;

import com.backend.BenchMarks.model.dto.ComparacaoDTO;
import com.backend.BenchMarks.model.enums.TipoLocalidade;
import com.fasterxml.jackson.annotation.JsonFormat;
import com.fasterxml.jackson.annotation.JsonManagedReference;

import jakarta.persistence.CascadeType;
import jakarta.persistence.Column;
import jakarta.persistence.Entity;
import jakarta.persistence.Enumerated;
import jakarta.persistence.EnumType;
import jakarta.persistence.GeneratedValue;
import jakarta.persistence.GenerationType;
import jakarta.persistence.Id;
import jakarta.persistence.OneToOne;
import jakarta.persistence.Table;
import jakarta.validation.constraints.NotNull;
import lombok.Data;

@Data
@Entity
@Table(name="benchmark")
public class BenchMark {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @NotNull(message = "O campo nome não pode ser vazio.")
    private String nome;

    private String localidade1;

    private String localidade2;

    @NotNull(message = "O campo estadoLocalidade2 não pode ser vazio.")
    String estadoLocalidade1;

    @NotNull(message = "O campo estadoLocalidade2 não pode ser vazio.")
    String estadoLocalidade2;

    @JsonFormat(pattern = "yyyy-MM-dd")
    @NotNull(message = "O campo data inicial não pode ser vazio.")
    private LocalDate dataInicial;
    
    @JsonFormat(pattern = "yyyy-MM-dd")
    @NotNull(message = "O campo data final não pode ser vazio.")
    private LocalDate dataFinal;

    @NotNull(message = "É obrigatório identificar qual o tipo de localidade.")
    @Enumerated(EnumType.STRING)
    TipoLocalidade tipoLocalidade;

    @JsonManagedReference
    @OneToOne(cascade = CascadeType.ALL, mappedBy = "benchmark", orphanRemoval = true)
    private ComparacaoDTO comparacao;
}

